dask/base.py

```diff
@@ -211,7 +211,11 @@ def is_dask_collection(x) -> bool:
 
     """
     try:
-        return x.__dask_graph__() is not None
+        if "xarray" in type(x).__module__:
+            return x.__dask_graph__() is not None
+        else:
+            return hasattr(x, "__dask_graph__")
```

Comment:
@phofl this change is actually kinda important.  We were generating the task graph before whenever calling `is_dask_collection`.  Maybe caching was saving us here, but this could be unpleasantly expensive for dask-expr.

The history here is that xarray collections have a `__dask_graph__` attribute, but are not always dask collections (they sometimes hold dask arrays but sometimes just have numpy arrays).

Reply:
Yeah I remember that we discussed this a while back and ended at revisiting if necessary because caching worked fine back then. That change is nice anyway.

Reply:
This seems pretty expensive for the xarray code path which really should just be something like
```
any(variable._data for variable in xarray_obj.variables)
```

Can we add something like `__dask_is_present__`?

Reply:
We can add a new protocol.  We'll have to do this everywhere, but it seems like an ok thing to do.  In the meantime, let's just special-case something for Xarray.  I think that it's the only project today where this matters.  

I started doing this:

```python
        if "xarray" in type(x).__module__:
            import xarray
            if isinstance(x, xarray.Dataset):
                return any(variable._data for variable in x.variables)
            if isinstance(x, xarray.DataArray):
                return bool(x._data)
```

But I'm not sure that `_data` exists?  Maybe we want `any(is_dask_collection(variable.data) ...)`?  Any suggestions?

Reply:
I suspect this is it:
```python
from dask.base import is_dask_collection


def is_dask_xarray(x):
    import xarray

    if isinstance(x, xarray.Dataset):
        return any(
            is_dask_collection(variable._data) for _, variable in x.variables.items()
        )
    elif isinstance(x, xarray.DataArray):
        return is_dask_collection(x.variable._data)
    elif isinstance(x, xarray.Variable):
        return is_dask_collection(x._data)


import xarray as xr

ds = xr.tutorial.open_dataset("air_temperature", chunks="auto")
computed = ds.compute()

assert is_dask_xarray(ds)
assert is_dask_xarray(ds.air)
assert is_dask_xarray(ds.air.variable)
assert not is_dask_xarray(computed)
assert not is_dask_xarray(computed.air)
assert not is_dask_xarray(computed.air.variable)
```

Reply:
I also ran into the issue of materialization on `__dask_graph__` when playing with the scheduler integration and I strongly recommend moving dask-expr away from this in favor of an explicit method doing that (that's what I am proposing in https://github.com/dask-contrib/dask-expr/pull/294)

Reply:
@dcherian I've pushed up your changes.  Thanks.  Using the `air_temperature` dataset requires an optional dependency `pooch`, which I was hesitant to add to CI just for this (although maybe I should if it's very lightweight).  Alternatively, can you recommend a way to construct a dataset using maybe `da.random.random`?

Reply:
@fjetter I'm not sure yet what you mean in the comment above.  We need a way to determine if an object is a dask collection.  I looked through the PR you mentioned and didn't immediately something there that would solve this problem.  Can you help me to understand?

Reply:
@mrocklin: `ds = xr.Dataset({"air": (("time", "lat", "lon"), dask.array.ones((120, 120, 120)))})`